servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/HttpMessageDiscardWatchdogClientFilter.java

```diff
@@ -0,0 +1,132 @@
+ * processing message payload bodies are cleaned up.
+ */
+final class HttpMessageDiscardWatchdogClientFilter implements StreamingHttpClientFilterFactory {
+
+    private static final ContextMap.Key<AtomicReference<Publisher<?>>> MESSAGE_PUBLISHER_KEY = ContextMap.Key
+            .newKey(HttpMessageDiscardWatchdogClientFilter.class.getName() + ".messagePublisher",
+                    generifyAtomicReference());
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(HttpMessageDiscardWatchdogClientFilter.class);
+
+    /**
+     * Instance of {@link HttpMessageDiscardWatchdogClientFilter}.
+     */
+    static final StreamingHttpClientFilterFactory INSTANCE = new HttpMessageDiscardWatchdogClientFilter();
+
+    /**
+     * Instance of {@link HttpLifecycleObserverRequesterFilter} with the cleaner implementation.
+     */
+    static final StreamingHttpClientFilterFactory CLEANER = new CleanerStreamingHttpClientFilterFactory();
+
+    private HttpMessageDiscardWatchdogClientFilter() {
+        // Singleton
+    }
+
+    @Override
+    public StreamingHttpClientFilter create(final FilterableStreamingHttpClient client) {
+        return new StreamingHttpClientFilter(client) {
+            @Override
+            protected Single<StreamingHttpResponse> request(final StreamingHttpRequester delegate,
+                                                            final StreamingHttpRequest request) {
+                return delegate.request(request).map(response -> {
+                    // always write the buffer publisher into the request context. When a downstream subscriber
+                    // arrives, mark the message as subscribed explicitly (having a message present and no
+                    // subscription is an indicator that it must be freed later on).
+                    final AtomicReference<Publisher<?>> reference = request.context()
+                            .computeIfAbsent(MESSAGE_PUBLISHER_KEY, key -> new AtomicReference<>());
+                    assert reference != null;
+                    final Publisher<?> previous = reference.getAndSet(response.messageBody());
+                    if (previous != null) {
+                        // If a previous message exists, the Single<StreamingHttpResponse> got resubscribed to
+                        // (i.e. during a retry) and so previous message body needs to be cleaned up.
+                        LOGGER.warn("Automatically draining previous HTTP response message body that was " +
+                                "not consumed. Users-defined retry logic must drain response payload before " +
+                                "retrying.");
+
+                        previous.ignoreElements().subscribe();
+                    }
+
+                    return response.transformMessageBody(msgPublisher -> msgPublisher.beforeSubscriber(() -> {
+                        final AtomicReference<?> maybePublisher = request.context().get(MESSAGE_PUBLISHER_KEY);
```

Comment:
Here and in `HttpMessageDiscardWatchdogServiceFilter`, can we reuse the above `reference` instead of querying `request.context()`? Doesn't look like we ever remove `MESSAGE_PUBLISHER_KEY`

Reply:
Good catch / simplification! 👍